src/main/java/redis/clients/jedis/executors/CircuitBreakerCommandExecutor.java

```diff
@@ -0,0 +1,95 @@
+
+    private final MultiClusterPooledConnectionProvider provider;
+
+    public CircuitBreakerCommandExecutor(MultiClusterPooledConnectionProvider provider) {
+        this.provider = provider;
+    }
+
+    @Override
+    public void close() {
+        IOUtils.closeQuietly(this.provider);
+    }
+
+    @Override
+    public <T> T executeCommand(CommandObject<T> commandObject) {
+        Cluster cluster = provider.getCluster(); // Pass this by reference for thread safety
+
+        DecorateSupplier<T> supplier = Decorators.ofSupplier(() -> this.handleExecuteCommand(commandObject, cluster));
+
+        supplier.withRetry(cluster.getRetry());
+        supplier.withCircuitBreaker(cluster.getCircuitBreaker());
+        supplier.withFallback(circuitBreakerFallbackException,
+                              e -> this.handleClusterFailover(commandObject, cluster.getCircuitBreaker()));
+
+        return supplier.decorate().get();
+    }
+
+    /**
+     * Functional interface wrapped in retry and circuit breaker logic to handle happy path scenarios
+     */
+    private <T> T handleExecuteCommand(CommandObject<T> commandObject, Cluster cluster) {
+        try (Connection connection = cluster.getConnection()) {
+            return connection.executeCommand(commandObject);
+        }
+    }
+
+    /**
+     * Functional interface wrapped in retry and circuit breaker logic to handle open circuit breaker failure scenarios
+     */
+    private synchronized <T> T handleClusterFailover(CommandObject<T> commandObject, CircuitBreaker circuitBreaker) {
+
+        // Check state to handle race conditions since incrementActiveMultiClusterIndex() is non-idempotent
+        if (!CircuitBreaker.State.FORCED_OPEN.equals(circuitBreaker.getState())) {
+
+            // Transitions state machine to a FORCED_OPEN state, stopping state transition, metrics and event publishing.
+            // To recover/transition from this forced state the user will need to manually failback
+            circuitBreaker.transitionToForcedOpenState();
+
+            // Incrementing the activeMultiClusterIndex will allow subsequent calls to the executeCommand()
+            // to use the next cluster's connection pool - according to the configuration's prioritization/order
+            int activeMultiClusterIndex = provider.incrementActiveMultiClusterIndex();
+
+            // Implementation is optionally provided during configuration. Typically, used for activeMultiClusterIndex persistence or custom logging
+            provider.runClusterFailoverPostProcessor(activeMultiClusterIndex);
+        }
+
+        // Once the priority list is exhausted only a manual failback can open the circuit breaker so all subsequent operations will fail
+        else if (provider.isLastClusterCircuitBreakerForcedOpen()) {
+            throw new JedisConnectionException("Cluster/database endpoint could not failover since the MultiClusterJedisClientConfig was not " +
```

Comment:
Is this a definite connection error? If so, it'd be a good idea to include that in this exception either as a cause or suppressed. Otherwise, use other exception.

Reply:
I'm not sure I understand your meaning. This condition is best handled with a checked exception so instead of creating yet another exception type I returned a connection exception because the problem is related to not being able to connect. I'm not sure what is at issue?